.changeset/little-ads-return.md

```diff
@@ -0,0 +1,106 @@
+previously paginated - in order to get all your pages, you may now use our
+`toArray` pagination helper method, which will automatically paginate through
+all results and aggregate them into an array:
+
+```tsx
+import { client } from '@/makeswift/client'
+import { MakeswiftPage } from '@makeswift/runtime/next'
+
+async function getAllPages(): Promise<MakeswiftPage[]> {
+  return await client.getPages().toArray()
+}
+```
+
+`getPages` now returns an instance of `IterablePaginationResult`, a decorated
+async iterator which includes methods `.map` and `.filter`, in addition to
+`.toArray`, mentioned above.
+
+This change also deprecates the client `getSitemap` method, with the
+recommendation that sitemaps should now be generated using data returned from
+`getPages`. Note that the deprecation of `getSitemap` now involves the host
+being responsible for the construction of page paths in the sitemap (either with
+domain or path based localization). Below is an example that uses path-based
+localization with the `next-sitemap` library:
+
+```tsx pages/sitemap.xml.tsx
+import { getServerSideSitemapLegacy } from 'next-sitemap'
+import { MakeswiftPage } from '@makeswift/runtime/next'
+import { client } from '@makeswift/client'
+
+const DOMAIN = 'https://example.com'
+const DEFAULT_PRIORITY = 0.75
+const DEFAULT_FREQUENCY = 'hourly'
+
+function pageToSitemapItem(page: MakeswiftPage) {
+  const pageUrl = new URL(page.path, DOMAIN)
+  return {
+    loc: pageUrl.href,
+    lastmod: page.createdAt,
+    changefreq: page.sitemapFrequency ?? DEFAULT_FREQUENCY,
+    priority: page.sitemapPriority ?? DEFAULT_PRIORITY,
+    alternateRefs: page.localizedVariants.map(variant => {
+      const localizedPath = `/${variant.locale}/${variant.path}`
+      const localizedPageUrl = new URL(localizedPath, DOMAIN)
+      return {
+        hreflang: variant.locale,
+        href: localizedPageUrl.href,
+      }
+    }),
+  }
+}
+
+export async function getServerSideProps(context) {
+  const sitemap = client
+    .getPages()
+    .filter(page => !page.excludedFromSearch)
+    .map(page => pageToSitemapItem(page))
+    .toArray()
+
+  return getServerSideSitemapLegacy(context, sitemap)
+}
+
+export default function Sitemap() {}
+
+```
+
+Here's another example for Next.js's App Router built-in support for sitemaps:
+
+```ts app/sitemap.ts
+
+import { MetadataRoute } from 'next'
+import { MakeswiftPage } from '@makeswift/runtime/dist/types/next'
+import { client } from '@/lib/makeswift/client'
+
+type NextSitemapItem = MetadataRoute.Sitemap[number]
+
+const DOMAIN = 'https://example.com'
+const DEFAULT_PRIORITY = 0.75
+const DEFAULT_FREQUENCY = 'hourly'
+
+function pageToSitemapEntry(page: MakeswiftPage): NextSitemapItem {
+  const pageUrl = new URL(page.path, DOMAIN)
+  return {
+    url: pageUrl.href,
+    lastModified: page.createdAt,
+    changeFrequency: page.sitemapFrequency ?? DEFAULT_FREQUENCY,
+    priority: page.sitemapPriority ?? DEFAULT_PRIORITY,
```

Comment:
What about the variants?

Reply:
This is the funniest thing, you currently cannot generate a localized sitemap. There's an issue on Github for it: https://github.com/vercel/next.js/issues/65817. To replicate this yourself:

1. Create a new Next.js app with `npx create-next-app@latest` (accept all the default settings)
2. Create a new `app/sitemap.ts` file
3. Copy and paste their example from: https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generating-a-sitemap-using-code-js-ts. This will work when you visit `localhost:3000/sitemap.xml`
4. Now, copy and paste their localized example: https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-localized-sitemap. This will not work. Remove all the `alternates`, and it works again.

I intentionally omitted this from the example for now.

Reply:
Got it, I think we should file a new issue for it since the linked one didn't follow their template and got auto-closed.

Reply:
Yeah, we can also reach out via Slack if needed.

Reply:
Opened an issue: https://github.com/vercel/next.js/issues/67005